group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -2288,18 +2259,40 @@ private CoordinatorResult<StreamsGroupHeartbeatResponseData, CoordinatorRecord>
             records
         );
 
-        if (bumpGroupEpoch || group.hasMetadataExpired(currentTimeMs)) {
-            // The subscription metadata is updated when the refresh deadline has been reached.
-            subscriptionMetadata = group.computeSubscriptionMetadata(
+        // 2. Initialize/Update the group topology.
```

Comment:
If those different steps are so important, why not exporting them to methods and give them meaningful names?

Reply:
I agree this method should be split. But splitting it now would make things even more confusing, because there are 4 different heartbeat implementations with similar sub-steps in this class. I know you don't like inline comments, but let's postpone splitting this method to when we have split the `GroupMetadataManager`.